### ISSUE TITLE
Show # unread messages in account switcher

### ISSUE THREAD

**Issue by jewel-andraia** [honestbleeps/Reddit-Enhancement-Suite#1428]
get the user's unread RSS feed from prefs: http://www.reddit.com/r/Enhancement/comments/2a4rsk/feature_request_inbox_totals_on_the_user_dropdown/

maybe there's even an api endpoint for requesting that

https://www.reddit.com/r/Enhancement/comments/3d55dz/feature_request_ui_tweak_show_inbox_message/

**Comment by andytuba** [comment]
more dupes https://www.reddit.com/r/Enhancement/comments/2sk8ew/feature_request_show_number_of_unread_messages_on/

**Comment by erikdesjardins** [comment]
More:
https://www.reddit.com/r/Enhancement/comments/2v8wjo/is_there_a_way_to_display_a_mail_notification_in/

**Comment by andytuba** [comment]
https://www.reddit.com/r/Enhancement/comments/2w5g6w/feature_request_send_connect_accounts_an_inbox/

a twist: orangered in main account

**Comment by erikdesjardins** [comment]
https://www.reddit.com/r/Enhancement/comments/31wbnr/feature_request_add_a_couple_of_simple_but_useful/
also: clicking on an account's orangered / mail count simultaneously logs you into that account and navigates to the inbox.
(I think that's a good idea)

**Comment by jeffzemla** [comment]
one cheap hack would be to keep track of all posts made by any account within the last 24-48 hours. by keeping a list of these, you could periodically check them for replies without switching accounts. this wouldn't catch everything, but it would catch most.

alternatively, a stripped down version of this feature could display the message counter as of when you were last logged in, for cases in which you don't clear your inbox.

**Comment by allthefoxes** [comment]
>rack of all posts made by any account within the last 24-48 hours.

Muh Storage Usage :fearful:

**Comment by jeffzemla** [comment]
It's just a handful of post ids. I realize this is not a very good
solution, but I don't think storage is really a concern.

**Comment by andytuba** [comment]
storage and API hits are both a concern. we'll do it right or not at all.  thanks for the suggestion, though!

**Comment by andytuba** [comment]
I'd really like to implement this by hitting the API with different OAuth tokens for each account... as soon as we build out OAuth authentication inside RES.

**Comment by TheLQ** [comment]
How about saving the cookies of each user? Then there isn't really too much overhead.

**Comment by andytuba** [comment]
Hmm. The authentication cookie is HttpOnly, which means client scripts can't access it. RES might be able to use the cookie API to fetch it anyway ... and then figure out how to select which cookie to send on the request.  Sounds about as hard as figuring out OAuth.

**Comment by andytuba** [comment]
Good pro/con ideas on https://www.reddit.com/r/Enhancement/comments/3d55dz/feature_request_ui_tweak_show_inbox_message/ct35twm?context=3

**Comment by andytuba** [comment]
Show unread badge on Acct switcher bhutton

**Comment by jsayol** [comment]
Just letting you guys know that I'm working on a possible solution to this. I'll keep you posted.

**Comment by andytuba** [comment]
@jsayol exciting!

**Comment by jsayol** [comment]
I've been giving a lot of thought [to this](https://www.reddit.com/r/Enhancement/comments/3d55dz/feature_request_ui_tweak_show_inbox_message/ct35twm?context=3). Here are some considerations:
* Feeds are enabled for everyone (even brand-new accounts, just checked) plus there's no way to disable/enable them. That simplifies this quite a bit as RES doesn't need to change any account settings to make this work.
* A feed id is no different than a password, so storing it in RES should not be a problem.

Taking that into account, here's the solution I'm going to implement (unless someone can give some strong reasons to do things differently):
* This functionality will be enabled by default, with a switch to disable it in the settings.
* Any new account added to RES will have this enabled, unless the user chooses otherwise when adding it.
* Any existing account in RES will have this disabled by default. This will prevent RES from accessing any account in situations where the user might not want to leave a trace (like accessing the user's nsfw account when they're at work, thus leaving an entry in the office server's logs... not good. Why would anyone add such an account to their workplace computer in the first place if they never plan on accessing it from there, though? Probably no one would, but better safe than sorry).
* Whenever a user logs into an account defined in the account switcher that has this feature enabled, RES will check if it has the feed URL for it. If not, it will grab it.
* Add 3 new fields to the account switcher settings table:
    * "switch" field to enable/disable this feature for that account.
    * "button" field to get/refresh the feed URL for that account.
    * "hidden" field (new type) to store the feed URL.
* Clicking the get/refresh button will temporarily log the user into that account, get the feed URL, and switch back to the original account. We don't want the user doing ANYTHING while this is going on (shouldn't take longer than 3 or 4 seconds) so we will add an overlay over the whole page preventing the user from performing any action. This overlay will have a short message explaining what is going on, something like: "Obtaining access to the account. Just a moment, please..."
* To make sure we properly deal with situations where the user has lots of accounts defined in the switcher, the procedure to check for unread messages will be as follows:
    * Instead of making AJAX requests straight away, add the account to a queue (unless it's already in it).
    * Take an account from the top of the queue and make the request. Do this every 5 seconds (from the moment the previous request finishes) until the queue is empty.
    * When the queue is empty, add all the accounts with this feature enabled to the queue. Wait 5 seconds, start processing it again.
    * Skip any account that has already been checked in the last 60 seconds, and move it to the bottom of the queue (we still want to check it, not just yet).
    * Skip the account for the currently logged-in user. You already have the notification up there, so this saves an unnecessary AJAX request.
    * Both the queue and the "last time checked" values are stored in memory, not in the persistent storage, so they reset any time the page reloads. This is done mostly to keep things simple.
* If a feed URL returns anything other than what we expect (probably because the user changed the password) we remove it from the settings. We obtain it again the next time the user logs into that account or clicks the relevant button in the settings (after updating the password for it).

I'm not going to start working on it just yet, I'll wait a couple of days to see if there are any comments or suggestions. Let me know what you think.

**Comment by jsayol** [comment]
I just realized something:

> Clicking the get/refresh button will temporarily log the user into that account, get the feed URL, **and switch back to the original account.**

That will only work in the original account is one of those defined in the account switcher and has an up-to-date password. Otherwise we won't be able to return to that account and will leave the user logged out. Any ideas?

**Comment by jsayol** [comment]
I haven't had much time lately for this but I'll get to it soon. Any ideas or comments?

**Comment by andytuba** [comment]
This seems like a fun and great thing to do, although a good chunk of work.  The "switch back and forth" 

We could get even more mileage out of it by replacing the "get the feed URL" step with "get OAuth authorization" ... although that means bringing in more knowledge to manage oauth tokens.  snoocore would probably help lots with that.

**Comment by erikdesjardins** [comment]
https://www.reddit.com/r/Enhancement/comments/40dut1/feature_request_a_button_to_check_every_alt/

**Comment by andytuba** [comment]
https://www.reddit.com/r/Enhancement/comments/4avqc8/feature_request_add_unread_message_indicator_to/

**Comment by erikdesjardins** [comment]
https://www.reddit.com/r/Enhancement/comments/4j2ms3/is_there_a_way_to_show_number_of_new_inbox/

**Comment by andytuba** [comment]
https://www.reddit.com/r/Enhancement/comments/4zmzd9/feature_request_the_account_dropdown_list_for/

**Comment by bluelightzero** [comment]
is there not a way for RES to have it's own private session/cookies so it can log in and do account actions without affecting your current browser session? In the same way a stand-alone app might connect through http api.

Seems it would solve pretty much all the problems with

**Comment by andytuba** [comment]
@bluelightzero Yes, that's why this issue is in the [OAuthed API access project](https://github.com/honestbleeps/Reddit-Enhancement-Suite/projects/4#card-121507) 😉
